### ISSUE TITLE
Sometimes fitting is SOOOO SLOW

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#125]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/125) by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

This is actually a serious issue, because it makes quickly analyzing datasets hard, I write code, and run, but it just gets stuck. My current code prints out the arguments to the fitter before running, so that I can record instances that are super slow.  I've been waiting for this particular fit to finish the entire time I've been putting together this, which included manually adding commas to this example.

I don't know if it matters, but this is on python 3.4.

This example takes at least 10 minutes on the bashful machine., but I'm not sure because it hasn't finished.

```
import mass
import numpy as np
fitter = mass.MnKAlphaFitter()
fitter.fit(np.array([36,  49,  39,  41,  46,  46,  42,  42,  46,  52,  51,  53,  54,  48,  58,  46,  57,  51,
  61,  68,  63,  68,  73,  79,  66,  84,  78,  94,  84,  84,  74,  85,  76,  81,  83,  84,
 100,  95,  93,  82,  74,  83,  93, 102,  98,  79, 100, 113,  95,  88, 104,  94,  95, 110,
 112,  81, 106, 104, 110,  97, 105,  95, 103,  97,  95, 103,  84,  97,  79,  85,  84,  87,
  80,  71,  77,  89,  83,  81,  59]),
np.array([ 5870.25,  5870.75,  5871.25,  5871.75,  5872.25,  5872.75,  5873.25,  5873.75,
  5874.25,  5874.75,  5875.25,  5875.75,  5876.25,  5876.75,  5877.25,  5877.75,
  5878.25,  5878.75,  5879.25,  5879.75,  5880.25,  5880.75,  5881.25,  5881.75,
  5882.25,  5882.75,  5883.25,  5883.75,  5884.25,  5884.75,  5885.25,  5885.75,
  5886.25,  5886.75,  5887.25,  5887.75,  5888.25,  5888.75,  5889.25,  5889.75,
  5890.25,  5890.75,  5891.25,  5891.75,  5892.25,  5892.75,  5893.25,  5893.75,
  5894.25,  5894.75,  5895.25,  5895.75,  5896.25,  5896.75,  5897.25,  5897.75,
  5898.25,  5898.75,  5899.25,  5899.75,  5900.25,  5900.75,  5901.25,  5901.75,
  5902.25,  5902.75,  5903.25,  5903.75,  5904.25,  5904.75,  5905.25,  5905.75,
  5906.25,  5906.75,  5907.25,  5907.75,  5908.25,  5908.75,  5909.25])
)
```

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

If we can't make this reliably fast, could we have a timeout as an argument, so that it fails if it runs for more than say 30 seconds?

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

One thing I noticed by running in ipython with %pdb is that in my case it is using `_smear_lowEtail` despite having the guess for tail fraction equal to zero, and tail is is held. `_smear_lowEtail` is being called with a value of `P_tailfrac` of 0.001.

Somewhat up the callstack is this line:
```python
> /home/tes/virt_envs/py34/lib/python3.4/site-packages/mass-0.6.0-py3.4-linux-x86_64.egg/mass/mathstat/fitting.py(388)_mrqcof()
    386 
    387         y_model = self.theory_function(params, self.x)
--> 388         dyda = (self.theory_gradient(params, self.x).T * dpdi_grad).T
    389         if y_model.min() <= 0:
    390             y_model[y_model <= 0] = 1e-10
```

Suggesting the the gradient function is possibly doing a finite difference, and is varying at least one held parameter.


```

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Investigating the theory that it is varying held parameters, I looked into how this is done. The step side for parameters is determined by calling `fitter.stepsize(params)`, and always returns finite values. So monkey patched in a version that returns zero for held paramters, and fitting the same data from the first post, returns almost instantly.

This reveals two separate issues:
  
 * The maximum possible worst case performance
 * an optimization for when parameters are held.

```python
import mass
import numpy as np
counts = np.array([36,  49,  39,  41,  46,  46,  42,  42,  46,  52,  51,  53,  54,  48,  58,  46,  57,  51,
  61,  68,  63,  68,  73,  79,  66,  84,  78,  94,  84,  84,  74,  85,  76,  81,  83,  84,
 100,  95,  93,  82,  74,  83,  93, 102,  98,  79, 100, 113,  95,  88, 104,  94,  95, 110,
 112,  81, 106, 104, 110,  97, 105,  95, 103,  97,  95, 103,  84,  97,  79,  85,  84,  87,
  80,  71,  77,  89,  83,  81,  59])
energies = np.array([ 5870.25,  5870.75,  5871.25,  5871.75,  5872.25,  5872.75,  5873.25,  5873.75,
  5874.25,  5874.75,  5875.25,  5875.75,  5876.25,  5876.75,  5877.25,  5877.75,
  5878.25,  5878.75,  5879.25,  5879.75,  5880.25,  5880.75,  5881.25,  5881.75,
  5882.25,  5882.75,  5883.25,  5883.75,  5884.25,  5884.75,  5885.25,  5885.75,
  5886.25,  5886.75,  5887.25,  5887.75,  5888.25,  5888.75,  5889.25,  5889.75,
  5890.25,  5890.75,  5891.25,  5891.75,  5892.25,  5892.75,  5893.25,  5893.75,
  5894.25,  5894.75,  5895.25,  5895.75,  5896.25,  5896.75,  5897.25,  5897.75,
  5898.25,  5898.75,  5899.25,  5899.75,  5900.25,  5900.75,  5901.25,  5901.75,
  5902.25,  5902.75,  5903.25,  5903.75,  5904.25,  5904.75,  5905.25,  5905.75,
  5906.25,  5906.75,  5907.25,  5907.75,  5908.25,  5908.75,  5909.25])
fitter = mass.MnKAlphaFitter()
guess_params = fitter.guess_starting_params(counts, energies)
for (meaning, ind) in fitter.param_meaning.items():
    print(meaning+" %g"%guess_params[ind])
def stepsize(params):
    print("OVERLOADED STEPSIZE")
    eps = np.array((1e-3, params[0]/1e5, 1e-3, params[3]/1e5, 1e-3, 1e-3, 1e-4, 1e-2))
    for i in fitter.hold:
        eps[i]=0
    print(eps)
    return eps
fitter.stepsize=stepsize
fitter.fit(counts, energies, guess_params, vary_tail=False)
```

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Concerning your note about held parameters: I wonder is this as simple as the gradient is (needlessly) doing estimates of the gradient even for the held parameters? Or is that what you were saying was the second of the "two separate issues"?

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Yeah I think it's two separate issues, because presumably one could want to fit to this data with the tail free to learn about tail fractions.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Fixes #125 and tests for it.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Removing milestone: v0.6 (automated comment)
